src/xml_utils.cpp

```diff
@@ -0,0 +1,1850 @@
+
+    output.mzhigh[i] = spec.attribute("highMz").as_double();
+    if (std::isnan(output.mzhigh[i])) output.mzhigh[i] = NA_REAL;
+
+    output.bpcmz[i] = spec.attribute("basePeakMz").as_double();
+    if (std::isnan(output.bpcmz[i])) output.bpcmz[i] = NA_REAL;
+
+    output.bpcint[i] = spec.attribute("basePeakIntensity").as_double();
+    if (std::isnan(output.bpcint[i])) output.bpcint[i] = NA_REAL;
+
+    output.ticint[i] = spec.attribute("totIonCurrent").as_double();
+    if (std::isnan(output.ticint[i])) output.ticint[i] = NA_REAL;
+
+    pugi::xml_node precursor = spec.child("precursorMz");
+
+    if (precursor != NULL) {
+      output.pre_mz[i] = precursor.text().as_double();
+      if (std::isnan(output.pre_mz[i])) output.pre_mz[i] = NA_REAL;
+
+      output.pre_ce[i] = spec.attribute("collisionEnergy").as_double();
+      if (std::isnan(output.pre_ce[i])) output.pre_ce[i] = NA_REAL;
+
+    } else {
+      output.pre_mz[i] = NA_REAL;
+      output.pre_ce[i] = NA_REAL;
+    }
+
+    output.pre_scan[i] = NA_INTEGER;
+    output.pre_loweroffset[i] = NA_REAL;
+    output.pre_upperoffset[i] = NA_REAL;
+  }
+}
+
+
+
+
+xml_utils::spectraHeaders xml_utils::parse_spectra_headers(const pugi::xml_node& root) {
+
+  xml_utils::spectraHeaders output;
+
+  const char* root_name = root.name();
+
+  if (strcmp("indexedmzML", root_name) == 0) {
+    xml_utils::mzml_spectra_headers_parser(root.child("mzML"), output);
+
+  } else if (strcmp("mzXML", root_name) == 0) {
+    xml_utils::mzxml_spectra_headers_parser(root, output);
+
+  } else {
+    std::cout << "\u2717 The file must be in valid mzML or mzXML format!" << std::endl;
+  }
+
+  return output;
+}
+
+
+
+
+Rcpp::List xml_utils::spectraHeaders_to_list(const xml_utils::spectraHeaders& headers_cpp) {
+
+  Rcpp::StringVector polarity = Rcpp::wrap(headers_cpp.polarity);
+  Rcpp::StringVector mode = Rcpp::wrap(headers_cpp.mode);
+
+  for (int i = 0; i < polarity.size(); i++) {
+    if (polarity(i) == "NA") {
+      polarity(i) = NA_STRING;
+    }
+    if (mode(i) == "NA") {
+      mode(i) = NA_STRING;
+    }
+  }
+
+  Rcpp::List headers;
+
+  headers["index"] = headers_cpp.index;
+  headers["scan"] = headers_cpp.scan;
+  headers["traces"] = headers_cpp.traces;
+  headers["polarity"] = polarity;
+  headers["mode"] = mode;
+  headers["level"] = headers_cpp.level;
+  headers["rt"] = headers_cpp.rt;
+  headers["drift"] = headers_cpp.drift;
+  headers["bpc_mz"] = headers_cpp.bpcmz;
+  headers["bpc_intensity"] = headers_cpp.bpcint;
+  headers["tic_intensity"] = headers_cpp.ticint;
+  headers["mzlow"] = headers_cpp.mzlow;
+  headers["mzhigh"] = headers_cpp.mzhigh;
+  headers["pre_scan"] = headers_cpp.pre_scan;
+  headers["pre_mz"] = headers_cpp.pre_mz;
+  headers["pre_ce"] = headers_cpp.pre_ce;
+  headers["pre_loweroffset"] = headers_cpp.pre_loweroffset;
+  headers["pre_upperoffset"] = headers_cpp.pre_upperoffset;
+
+  headers.attr("class") = Rcpp::CharacterVector::create("data.table", "data.frame");
```

Comment:
Is this a 'safe' way to create DTs from C++? (I always did the conversion in R).